Migration to a new framework for doc testing

niwatorrri opened the pull request:
This PR intends to create a new framework for testing all the current docs (that need to be tested). The old framework is still kept for transition.


[ ] Configuration Infrastructure
[x] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[x] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
Hi @niwatorrri. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio.io).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

niwatorrri commented:
@davidhauck Please take a look

shamsher31 commented:
/ok-to-test

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio.io | 73942759fe15a1005ade3de730e0d7291df67dde | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/7465/gencheck_istio.io/1097) | `/test gencheck_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

frankbu commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+
+	snipsFileSuffix = "/snips.sh"
+	testFileSuffix  = "/test.sh"
+	testCleanupSep  = "#! cleanup"
```
```suggestion
	testCleanupSep  = "# !cleanup"
```
Don't use `#!` since it's a very special Unix character sequence called `shebang`. From Wikipedia:
```
The shebang is actually a human-readable instance of a magic number in the executable file, the magic byte
string being 0x23 0x21, the two-character encoding in ASCII of #!. This magic number is detected by the "exec"
family of functions, which determine whether a file is a script or an executable binary. The presence of the
shebang will result in the execution of the specified executable, usually an interpreter for the script's language. 
```

niwatorrri replied:
How about `### cleanup ###`? This might be even less confusing

frankbu replied:
I think that looks like it could just be an ordinary comment, which it's not. I like David's suggestion `# @cleanup`. That makes it clear it's something special (for the framework), but not to be confused with something special for the bash interpreter.

niwatorrri replied:
Sounds good! I'll go with `# @cleanup` then

frankbu commented on tests/README.md:
```diff
@@ -214,16 +179,22 @@ make test.kube.trafficmanagement
 1. If HUB and TAG aren't set, then their default values will match what is used by the prow tests.
 
-## Running Tests: go test
+## Migrate from Old Framework
```
I don't think we want to document migration instructions. As soon as the new framework is ready, we can do a one-time conversion of the existing 22 tests in a single PR, and also in the same PR remove the old framework code.

niwatorrri replied:
Sounds good! I'll remove this part

frankbu commented on tests/README.md:
```diff
@@ -54,59 +56,13 @@ To write an `istio.io` test, follow these steps:
+Your bash script will consist of a series of test steps that call the commands in your generated `snips.sh` file, as well as a series of cleanup steps that should be run after everything is done.
 
-Your test script can then invoke the commands by simply calling snip functions:
+NOTE (for old framework authors): your script doesn't need to include the `snips.sh` or `util/samples.sh` any more. They will be automatically sourced and you can directly call any function defined in them:
```
Auto source `snips.sh` is fine, but I don't think we want to do that with `util/samples.sh` because many tests don't need it. In fact, as we write more tests, we may find that there are other common utils that are only needed by a different set of tests (e.g., utils shared by several observability tests) that will be in a similar optional include file (e.g., util/observe.sh). Those kinds of helper functions should not be auto sourced.

niwatorrri replied:
Sounds reasonable. I'll leave this sourcing up to the test authors. What about `util/verify.sh`? I think sourcing this should be fine.

niwatorrri replied:
A caveat would be that sourcing needs to be done for both test and cleanup: they are now placed in one single file but are actually run in different shells I believe.

frankbu replied:
There are 4 files that can be auto sourced (since everybody needs them).

1. util/verify.sh
2. util/debug.sh
3. util/helpers.sh
4. snip.sh

Anything else is optional, so it should be explicitly sourced by the test author, instead.

niwatorrri replied:
Gotcha! I've made the changes accordingly

davidhauck commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+
+func split(testsAsString string) []string {
+	testsAsSlice := strings.Split(testsAsString, ",")
+	for i := 0; i < len(testsAsSlice); i++ {
```
Go has a foreach loop that makes things slightly easier to read: https://golangcode.com/foreach-loops/

niwatorrri replied:
Changed and it looks better!

davidhauck commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+	return testsAsSlice
+}
+
+// setup for all tests
```
We can still use one testmain, but we'll have to set up the environment differently depending on the test. The go logic can handle the setup/teardown of the different framework instances

frankbu replied:
Yes, my suggestion is to add a `# @setup profile=<name>` comment at the top of the script. The good news is that most docs assume the default or demo profile is used or, for setup docs, Istio isn't even installed. So, at least to start, I think all we would need to support are 3 possible setups:

1. `# @setup profile=default`
2. `# @setup profile=demo`
3. `# @setup profile=none`

We'll need to extend this later, when we start testing multicluster docs and such.

niwatorrri replied:
It is possible that the tests to be run require different setups, which could require different profiles to be installed and used on the same cluster (using different namespaces?). This should perhaps be implemented in the subtest but I'm not sure what changes I should make to the istio test framework part.

frankbu replied:
The Istio setup on a cluster is for all namespaces. Most tests will, regardless of what namespace(s) they use, have some expectation of what Istio setup is done before the various commands run. I've noticed that most expect that it's either the default profile, demo profile, or no Istio (just start with a kube cluster). Some of these tests might change the configuration of Istio during their steps, but in their cleanup they reverse those things, so by the end of the test, it should be back to whatever it started with. So you really just need to worry about setting up the 3 configs that I mentioned, and run each test under one of those three mains. Your current TestMain implementation is the "default" profile case (btw, most of the current tests use that). Here's an example of one that uses the "demo" profile: https://github.com/istio/istio.io/blob/master/tests/examples/main_test.go. The "none" profile case looks like this: https://github.com/istio/istio.io/blob/4e5d2ea8a744752b9454e1bc79ac967d9e4dcc93/tests/security/authz_td_migration/main_test.go

frankbu commented on tests/README.md:
````diff
@@ -182,27 +138,36 @@ expected output. The framework includes the following built-in verify functions:
+```bash
+make doc.test TEST=traffic-management
+```
+will run all the tests under `traffic-management` folder. The `TEST` variable also accepts multiple test names separated by commas, for example,
````
Now that the tests are in the content tree, there may be more folders with the same name (e.g., under /docs/tasks and under /docs/ops), so you will need to qualify the names (e.g., TEST=tasks/traffic-management)

niwatorrri replied:
Sounds good! I'll document that. I didn't see snips.sh files under docs/ops so I thought it might not be a problem as the framework would only find `test.sh` files

frankbu commented on tests/README.md:
````diff
@@ -182,27 +138,36 @@ expected output. The framework includes the following built-in verify functions:
+will run all the tests under `traffic-management` folder. The `TEST` variable also accepts multiple test names separated by commas, for example,
 ```bash
-make test.kube.trafficmanagement
+make doc.test TEST=request-routing,fault-injection
````
Same comment as above ... need to qualify tests (TEST=tasks/traffic-management/request-routing,tasks/traffic-management/fault-injection)

frankbu commented:
@niwatorrri The bash tracing that I mentioned works like this in the current framework:

1. A util file named [debug.sh](https://github.com/istio/istio.io/blob/master/tests/util/debug.sh)  is automatically sourced in every test file. It does some bash magic with FD 5, which turns on bash tracing.
1. As a result, a file with the same name as the script, with `_debug.txt` appended, is created in the `out` directory, providing a full trace of the bash execution. For example, for the tcp traffic shifting test, the file `tcp_trafficshifting.sh_debug.txt` looks like this:

```
+set -e
+set -u
+set -o pipefail
+source /work/content/en/docs/tasks/traffic-management/tcp-traffic-shifting/snips.sh
++source(): read -r -d '' snip_apply_weightbased_tcp_routing_2_out
++source(): read -r -d '' snip_apply_weightbased_tcp_routing_4_out
++source(): read -r -d '' snip_apply_weightbased_tcp_routing_5_out
+source /work/tests/util/samples.sh
+kubectl -n istio-system patch service istio-ingressgateway --patch '
spec:
  ports:
    - port: 31400
      targetPort: 31400
      name: tcp
'
+snip_set_up_the_test_environment_1
+snip_set_up_the_test_environment_1(): kubectl create namespace istio-io-tcp-traffic-shifting
+snip_set_up_the_test_environment_1(): kubectl label namespace istio-io-tcp-traffic-shifting istio-injection=enabled
+snip_set_up_the_test_environment_2
+snip_set_up_the_test_environment_2(): kubectl apply -f samples/sleep/sleep.yaml -n istio-io-tcp-traffic-shifting
+snip_set_up_the_test_environment_3
+snip_set_up_the_test_environment_3(): kubectl apply -f samples/tcp-echo/tcp-echo-services.yaml -n istio-io-tcp-traffic-shifting
+sample_wait_for_deployment istio-io-tcp-traffic-shifting tcp-echo-v1
+sample_wait_for_deployment(): local namespace=istio-io-tcp-traffic-shifting
+sample_wait_for_deployment(): local name=tcp-echo-v1
+sample_wait_for_deployment(): kubectl -n istio-io-tcp-traffic-shifting rollout status deployment tcp-echo-v1 --timeout 5m
+sample_wait_for_deployment istio-io-tcp-traffic-shifting tcp-echo-v2
+sample_wait_for_deployment(): local namespace=istio-io-tcp-traffic-shifting
+sample_wait_for_deployment(): local name=tcp-echo-v2
+sample_wait_for_deployment(): kubectl -n istio-io-tcp-traffic-shifting rollout status deployment tcp-echo-v2 --timeout 5m
+sample_wait_for_deployment istio-io-tcp-traffic-shifting sleep
+sample_wait_for_deployment(): local namespace=istio-io-tcp-traffic-shifting
+sample_wait_for_deployment(): local name=sleep
+sample_wait_for_deployment(): kubectl -n istio-io-tcp-traffic-shifting rollout status deployment sleep --timeout 5m
+sample_set_ingress_environment_variables
++sample_set_ingress_environment_variables(): kubectl get svc istio-ingressgateway -n istio-system
+sample_set_ingress_environment_variables(): local 'extlb=NAME                   TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                                                                      AGE
istio-ingressgateway   LoadBalancer   10.100.228.142   <pending>     31400:30839/TCP,15021:31094/TCP,80:31323/TCP,443:30446/TCP,15443:30836/TCP   4m57s'
+sample_set_ingress_environment_variables(): [[ NAME                   TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                                                                      AGE
istio-ingressgateway   LoadBalancer   10.100.228.142   <pending>     31400:30839/TCP,15021:31094/TCP,80:31323/TCP,443:30446/TCP,15443:30836/TCP   4m57s != *\<\n\o\n\e\>* ]]
+sample_set_ingress_environment_variables(): [[ NAME                   TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                                                                      AGE
istio-ingressgateway   LoadBalancer   10.100.228.142   <pending>     31400:30839/TCP,15021:31094/TCP,80:31323/TCP,443:30446/TCP,15443:30836/TCP   4m57s != *\<\p\e\n\d\i\n\g\>* ]]
++sample_set_ingress_environment_variables(): kubectl get po -l istio=ingressgateway -n istio-system -o 'jsonpath={.items[0].status.hostIP}'
+sample_set_ingress_environment_variables(): export INGRESS_HOST=172.18.0.2
+sample_set_ingress_environment_variables(): INGRESS_HOST=172.18.0.2
++sample_set_ingress_environment_variables(): kubectl -n istio-system get service istio-ingressgateway -o 'jsonpath={.spec.ports[?(@.name=="http2")].nodePort}'
+sample_set_ingress_environment_variables(): export INGRESS_PORT=31323
+sample_set_ingress_environment_variables(): INGRESS_PORT=31323
++sample_set_ingress_environment_variables(): kubectl -n istio-system get service istio-ingressgateway -o 'jsonpath={.spec.ports[?(@.name=="https")].nodePort}'
+sample_set_ingress_environment_variables(): export SECURE_INGRESS_PORT=30446
+sample_set_ingress_environment_variables(): SECURE_INGRESS_PORT=30446
++sample_set_ingress_environment_variables(): kubectl -n istio-system get service istio-ingressgateway -o 'jsonpath={.spec.ports[?(@.name=="tcp")].nodePort}'
+sample_set_ingress_environment_variables(): export TCP_INGRESS_PORT=30839
+sample_set_ingress_environment_variables(): TCP_INGRESS_PORT=30839
+snip_apply_weightbased_tcp_routing_1
+snip_apply_weightbased_tcp_routing_1(): kubectl apply -f samples/tcp-echo/tcp-echo-all-v1.yaml -n istio-io-tcp-traffic-shifting
++snip_apply_weightbased_tcp_routing_2
++snip_apply_weightbased_tcp_routing_2(): for i in {1..20}
+++snip_apply_weightbased_tcp_routing_2(): kubectl get pod -l app=sleep -n istio-io-tcp-traffic-shifting -o 'jsonpath={.items..metadata.name}'
++snip_apply_weightbased_tcp_routing_2(): kubectl exec sleep-8f795f47d-vh9x4 -c sleep -n istio-io-tcp-traffic-shifting -- sh -c '(date; sleep 1) | nc 172.18.0.2 30839'
++snip_apply_weightbased_tcp_routing_2(): for i in {1..20}
+++snip_apply_weightbased_tcp_routing_2(): kubectl get pod -l app=sleep -n istio-io-tcp-traffic-shifting -o 'jsonpath={.items..metadata.name}'
++snip_apply_weightbased_tcp_routing_2(): kubectl exec sleep-8f795f47d-vh9x4 -c sleep -n istio-io-tcp-traffic-shifting -- sh -c '(date; sleep 1) | nc 172.18.0.2 30839'
++snip_apply_weightbased_tcp_routing_2(): for i in {1..20}
+++snip_apply_weightbased_tcp_routing_2(): kubectl get pod -l app=sleep -n istio-io-tcp-traffic-shifting -o 'jsonpath={.items..metadata.name}'
++snip_apply_weightbased_tcp_routing_2(): kubectl exec sleep-8f795f47d-vh9x4 -c sleep -n istio-io-tcp-traffic-shifting -- sh -c '(date; sleep 1) | nc 172.18.0.2 30839'
++snip_apply_weightbased_tcp_routing_2(): for i in {1..20}
+++snip_apply_weightbased_tcp_routing_2(): kubectl get pod -l app=sleep -n istio-io-tcp-traffic-shifting -o 'jsonpath={.items..metadata.name}'
++snip_apply_weightbased_tcp_routing_2(): kubectl exec sleep-8f795f47d-vh9x4 -c sleep -n istio-io-tcp-traffic-shifting -- sh -c '(date; sleep 1) | nc 172.18.0.2 30839'
...
```

niwatorrri commented on Makefile.core.mk:
```diff
@@ -142,8 +142,8 @@ foo2:
-	@echo "ISTIO_SHA = ${ISTIO_SHA}"
-	@echo "HUB = ${HUB}"
+	@echo "ISTIO_SHA=${ISTIO_SHA}"
+	@echo "HUB=${HUB}"
```
Ok sure. It's just my OCD happening lol

niwatorrri commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+// Copyright 2020 Istio Authors
```
Oh I thought it was like Java where the package name should be the same as the folder name. I'll change that. Maybe we can move these under tests/ as well?

niwatorrri replied:
I've moved this code under tests/ and renamed the package as framework, which sounds kind of parallel to "istio.io/istio/pkg/test/framework"

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio.io | 5b059dfb35520a5fcd8235d829deb0a25d83fd29 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/7465/gencheck_istio.io/1102) | `/test gencheck_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio.io | 982b54d003b45423e4091e1129755c94f0d3239e | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/7465/gencheck_istio.io/1112) | `/test gencheck_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+	testCleanupSep  = "#! cleanup"
+)
+
+func split(testsAsString string) []string {
```
I've added docs for all functions. Not sure if it looks good as I was expecting something like `@input xxx @return xxx` but the comments in the link didn't seem to work this way

niwatorrri commented on content/doc_test.go:
```diff
@@ -0,0 +1,154 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package content
```
Yup this is now moved by the new commits

niwatorrri commented:
FYI I made some changes regarding timeout:
1. added a TIMEOUT variable for `make doc.test`. Default is 30 minutes.
2. added two TODOs in the go code on timeout for individual subtests and parallelizing subtests.

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | 832db12272ba0d39ab6aaa0f474b48047fa0f654 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/7465/lint_istio.io/6990) | `/test lint_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
> @niwatorrri The bash tracing that I mentioned works like this in the current framework:
> 
> 1. A util file named [debug.sh](https://github.com/istio/istio.io/blob/master/tests/util/debug.sh)  is automatically sourced in every test file. It does some bash magic with FD 5, which turns on bash tracing.
> 2. As a result, a file with the same name as the script, with `_debug.txt` appended, is created in the `out` directory, providing a full trace of the bash execution. 

This now works with the new framework enabled by the commit above. Running `make doc.test TEST=request_routing` will generate two files: `out/content_en_docs_tasks_traffic-management_request-routing_test_debug.txt` and `out/content_en_docs_tasks_traffic-management_request-routing_cleanup_debug.txt`.

frankbu commented on tests/README.md:
````diff
@@ -182,27 +141,45 @@ expected output. The framework includes the following built-in verify functions:
+```sh
+# @cleanup
+```
+as a single line followed by the cleanup steps that clean up all the resources. These steps can also directly call functions defined in the auto-sourced scripts as described before. But others like `util/samples.sh` need to be sourced again in cleanup, even if you have already sourced them for the test steps.
````
> But others like `util/samples.sh` need to be sourced again in cleanup

That seems kind of ugly. Won't the bash linter complain about that? How about the framework look for any `source` lines in the main script, and put a copy in the cleanup script?

frankbu replied:
Another sort of related thought ... would it make sense to leave the cleanup steps in the main bash script, in addition to putting a copy in the cleanup script? Then in the "successful completion" (common) case, the test is completely finished after the main script completes. Only in the failure case, would you have to run the cleanup script. That would seem nicer to me (and more efficient, since you won't need to re-source all the includes).

niwatorrri replied:
Personally I'd prefer the first approach. The cleanup steps will be run anyway as they are `defer`-ed, and re-sourcing files are only minimal overhead compared to test themselves.

frankbu replied:
If you leave the cleanup steps in the main script, you would not `defer` the cleanup script. Instead you would just run the main script, if success you're done, if failure you run the cleanup script. I'm not sure, though, if the underlying istio test framework gives you the hooks to do that, but if it does, that does seem nicer to me.

That said, it's not a big deal, so whatever you decide is fine.

niwatorrri replied:
This is now resolved with the first solution as the istio test framework doesn't seem to allow that much flexibility.

frankbu commented:
> This now works with the new framework enabled by the commit above. Running make doc.test TEST=request_routing will generate two files: out/content_en_docs_tasks_traffic-management_request-routing_test_debug.txt and out/content_en_docs_tasks_traffic-management_request-routing_cleanup_debug.txt.

Every file will have the prefix `content_en_docs_`, so you could shorten the file names to leave that part off.

istio-testing commented:
@niwatorrri: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio.io | ac5523a1f952c905e1248f64f91d688c99f6cfa6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio.io/7465/lint_istio.io/6997) | `/test lint_istio.io`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

niwatorrri commented:
To support multiple istio setup configs, I have made a lot of changes to the structure of the code and the framework. It would be good if everyone could do another review.

Original `doc_test.go` is now `doc_test_framework.go`, which means this is not the main entry point any more. The entry point is now in `tests/setup/*/doc_test.go` where each folder corresponds to a setup config. Each doc_test.go will setup istio in its own TestMain and then call TestDocs in doc_test_framework.go. Readme is also updated regarding how to add other configs. I adopted this solution b/c `framework.Run()` will call `os.Exit()` directly after it's finished so different runs have to be placed in different packages.

To make the framework smarter, the flow is now as follows: `make doc.test` will run `go test tests/setup/...` to find doc_test.go for every setup config -> doc_test.go imports the framework, where info about each test is collected in its init() -> check if any test requires this setup -> if no then continue; if yes doc_test.go does the setup and calls TestDocs in the framework to run the tests. This is however suboptimal in that test info is collected every time a setup config is found. If necessary I could first collect the test info into a, say, json file before actually running tests.

The new framework now works like this (where three tests are run and I changed one setup profile to demo on purpose). I believe there are some changes in 1.6.1 which makes the log look kind of different from previous.

```
2020-06-06T23:39:04.706855Z     info    Starting test doc(s): all docs
2020-06-06T23:39:04.789855Z     info    Setting up istio with profile=default
2020-06-06T23:39:04.789855Z     info    tf      Test run dir: /tmp/profile-default-0344e8074032414

- Processing resources for Istio core.
✔ Istio core installed
- Processing resources for Istiod.
- Processing resources for Istiod. Waiting for Deployment/istio-system/istiod
✔ Istiod installed
- Processing resources for Ingress gateways.
- Processing resources for Ingress gateways. Waiting for Deployment/istio-system/istio-ingressgateway
✔ Ingress gateways installed
- Pruning removed resources
✔ Installation complete=== RUN   TestDocs
=== RUN   TestDocs/en/docs/tasks/traffic-management/fault-injection/test.sh
=== RUN   TestDocs/en/docs/tasks/traffic-management/request-routing/test.sh
--- PASS: TestDocs (306.64s)
    --- PASS: TestDocs/en/docs/tasks/traffic-management/fault-injection/test.sh (163.57s)
    --- PASS: TestDocs/en/docs/tasks/traffic-management/request-routing/test.sh (143.07s)
PASS
2020-06-06T23:45:28.556361Z     error   tf      Error during close: 1 error occurred:
        * configmaps "istio-status-leader" not found


ok      istio.io/istio.io/tests/setup/profile_default   383.930s
2020-06-06T23:45:34.782431Z     info    Starting test doc(s): all docs
2020-06-06T23:45:34.849524Z     info    Setting up istio with profile=demo
2020-06-06T23:45:34.850394Z     info    tf      Test run dir: /tmp/profile-demo-39f10f5c6c6f4ed9b0

- Processing resources for Istio core.
✔ Istio core installed
- Processing resources for Istiod.
- Processing resources for Istiod. Waiting for Deployment/istio-system/istiod
✔ Istiod installed
- Processing resources for Egress gateways, Ingress gateways.
- Processing resources for Addons, Egress gateways, Ingress gateways.
- Processing resources for Addons, Egress gateways, Ingress gateways. Waiting for Deployment/istio-system/istio-egressgateway
- Processing resources for Addons, Egress gateways, Ingress gateways. Waiting for Deployment/istio-system/istio-egressgateway, Deployment/istio-system...
- Processing resources for Addons, Egress gateways, Ingress gateways. Waiting for Deployment/istio-system/grafana, Deployment/istio-system/istio-egres...
✔ Ingress gateways installed
- Processing resources for Addons, Egress gateways. Waiting for Deployment/istio-system/grafana, Deployment/istio-system/istio-egressgateway, Deployme...
✔ Egress gateways installed
- Processing resources for Addons. Waiting for Deployment/istio-system/grafana, Deployment/istio-system/kiali
- Processing resources for Addons. Waiting for Deployment/istio-system/kiali
✔ Addons installed
- Pruning removed resources
✔ Installation complete=== RUN   TestDocs
=== RUN   TestDocs/en/docs/tasks/traffic-management/tcp-traffic-shifting/test.sh
--- PASS: TestDocs (180.90s)
    --- PASS: TestDocs/en/docs/tasks/traffic-management/tcp-traffic-shifting/test.sh (180.90s)
PASS
2020-06-06T23:50:04.547855Z     error   tf      Error during close: 1 error occurred:
        * configmaps "istio-status-leader" not found


ok      istio.io/istio.io/tests/setup/profile_demo      269.844s
2020-06-06T23:50:10.527855Z     info    Starting test doc(s): all docs
2020-06-06T23:50:10.593855Z     info    No tests need to be run with profile=none
ok      istio.io/istio.io/tests/setup/profile_none      0.138s
```

niwatorrri commented:
> > This now works with the new framework enabled by the commit above. Running make doc.test TEST=request_routing will generate two files: out/content_en_docs_tasks_traffic-management_request-routing_test_debug.txt and out/content_en_docs_tasks_traffic-management_request-routing_cleanup_debug.txt.
> 
> Every file will have the prefix `content_en_docs_`, so you could shorten the file names to leave that part off.

Are we assuming only English docs will be tested? If that's the case yeah I can make them short.

davidhauck commented on tests/doc_test_framework.go:
```diff
@@ -0,0 +1,255 @@
+	// read the script file
+	script, err := ioutil.ReadFile(path)
+	if err != nil {
+		return &TestCase{path: shortPath, err: err}
```
Usually in go, the error is returned as a second argument in the function instead of as a property of the returned object

niwatorrri replied:
Gotcha. Changed in the commit below.

davidhauck commented on tests/doc_test_framework.go:
```diff
@@ -0,0 +1,255 @@
+// TestDocs traverses through all test cases and runs those that need the
+// setup config specified by the input. The (*testing.T) variable comes
+// from the TestDocs function in each tests/setup/*/doc_test.go
+func TestDocs(t *testing.T, config string) {
```
PascalCase means that a method is public and camelCase means it's private. Let's make sure the functions have the correct casing.

niwatorrri replied:
These actually have to be public as NeedSetup and TestDocs are called from, say, `tests/setup/profile_default/doc_test.go`, so I'd assume this is okay.

nmittler commented on content/en/docs/tasks/traffic-management/traffic-shifting/test.sh:
```diff
@@ -0,0 +1,102 @@
+INGRESS_URL="http://istio-ingressgateway.istio-system"
+# reviews_v3_traffic_percentage
+# gets the % of productpage requests with reviews from reviews:v3 service
+function reviews_v3_traffic_percentage() {
```
nit: we should probably make this a general-purpose tool in the future. Maybe add a TODO?

niwatorrri replied:
ok I added a TODO there, though I didn't write that part of script actually

nmittler commented on tests/README.md:
```diff
@@ -11,7 +13,7 @@ This means that we extract and test the exact same commands that are published i
 around the [Istio test framework](https://github.com/istio/istio/wiki/Istio-Test-Framework).
 
-Run the following command to see the current test coverage, including the list of documents
+[UPDATE NEEDED] Run the following command to see the current test coverage, including the list of documents
```
Should we add TODOs and reference tracking bugs?

niwatorrri replied:
What steps does "reference tracking bugs" translate to? I'm not sure if there's an inventory of TODOs to keep them.

nmittler replied:
Ah right ... Usually we do something like `TODO(https://github.com/istio/istio/issues/<number>`. It just helps readers know that this work is scheduled and gives them a way to track the progress.

nmittler commented on tests/README.md:
```diff
@@ -54,62 +56,25 @@ To write an `istio.io` test, follow these steps:
-    See https://docs.google.com/document/d/1r_NoxatNjzPsw0eXr6_9W0rqlkfelt7QfN2y5TF0yTo/edit#.
-
-1. Create your test bash script in the `scripts/` subdirectory.
+1. Create your test bash script `test.sh` next to the `snips.sh` you have just generated. Scripts that have other names will be ignored.
```
@frankbu Is `make snips` automatically called from `make gen`? If not it should be. The `gen` target is used by CI to determine if all generated files are up-to-date.

frankbu replied:
Yes it is.

frankbu commented on content/en/docs/tasks/traffic-management/traffic-shifting/test.sh:
```diff
@@ -0,0 +1,103 @@
+INGRESS_URL="http://istio-ingressgateway.istio-system"
+# reviews_v3_traffic_percentage
+# gets the % of productpage requests with reviews from reviews:v3 service
+# TODO: maybe make this a general purpose tool
```
```suggestion
# TODO: generalize this function and move to samples.sh so it can be reused by
#       other tests that check traffic distribution.
```

frankbu commented on tests/README.md:
````diff
@@ -15,6 +17,7 @@ Run the following command to see the current test coverage, including the list o
 that are in need of a test:
 
 ```sh
+# TODO(https://github.com/istio/istio.io/issues/7508): update for the new test framework
````
The `make test_status` command is unaffected by this framework change.
```suggestion
```

niwatorrri replied:
Oh I added this because I saw this [comment](https://docs.google.com/document/d/1r_NoxatNjzPsw0eXr6_9W0rqlkfelt7QfN2y5TF0yTo/edit?disco=AAAAJlQR_W4). I thought you might want to determine whether a doc has a test by checking if there's a test.sh next to snips.sh

frankbu replied:
That will get even more confusing if there are mutliple test.sh files (so the doc might be partially tested). I think we should stick to the rule that if a PR adds a snip.sh file, it also MUST include the corresponding test, at least for now. We can reconsider in the future, if it doesn't work out well.

frankbu commented on tests/README.md:
```diff
@@ -54,62 +57,25 @@ To write an `istio.io` test, follow these steps:
-    See https://docs.google.com/document/d/1r_NoxatNjzPsw0eXr6_9W0rqlkfelt7QfN2y5TF0yTo/edit#.
-
-1. Create your test bash script in the `scripts/` subdirectory.
+1. Create your test bash script `test.sh` next to the `snips.sh` you have just generated. Scripts that have other names will be ignored.
```
#7501 is a WIP PR that is using making 3 different tests for the same doc/snip.sh. I'm not sure if that's a good idea (still need to see where that's going, and review), but if we decide it reasonable, maybe you should switch this back to running bash scripts with the suffix `test.sh` (your original design), instead of exactly `test.sh`. Most tests would still be just named `test.sh` but if somebody want's to split it into more than one bash script, they can have something like `part1_test.sh` and `part2_test.sh` in the same directory, and the framework will run both.

niwatorrri replied:
Sounds reasonable. I'm allowing flexibility on the script name. How about stating it like this:
```suggestion
1. Create your test bash script `test.sh` next to the `snips.sh` you have just generated. The bash script name(s) should only have `test.sh` as their suffix. Scripts with other suffixes will be ignored.
```

frankbu replied:
Here's my suggestion:
```
1. Create a test bash script named `test.sh` next to the `snips.sh` you have just generated.
   If your document is very large and you want to break it into multiple tests, create multiple scripts with the suffix `test.sh`, instead.
   Other scripts in the directory will be ignored.
```

frankbu commented on tests/README.md:
```diff
@@ -54,62 +57,25 @@ To write an `istio.io` test, follow these steps:
-a test for the traffic-shifting task will have the following line in the script:
+Your bash script will consist of a series of test steps that call the commands in your generated `snips.sh` file, as well as a series of cleanup steps that should be run after everything is done.
 
+To begin with the test steps, there should always be one line that specifies the istio setup configuration for the test. The setup line should take the form of
```
```suggestion
Before the test steps, there should be one line that specifies the istio setup configuration for the test. The setup line should take the form of
```

frankbu commented on tests/README.md:
````diff
@@ -54,62 +57,25 @@ To write an `istio.io` test, follow these steps:
-source "${REPO_ROOT}/content/en/docs/tasks/traffic-management/traffic-shifting/snips.sh"
+# @setup <setup_config>
 ```
+Currently supported setup configurations include: `profile=default` for installing default profile, `profile=demo` for installing demo profile, and `profile=none` for not installing istio at all. This line can be placed anywhere before the cleanup steps.
````
```suggestion
Currently supported setup configurations include: `profile=default` to install the default profile, `profile=demo` to install the demo profile, and `profile=none` to not installing istio at all.
```

frankbu commented on tests/README.md:
```diff
@@ -54,62 +57,25 @@ To write an `istio.io` test, follow these steps:
+Currently supported setup configurations include: `profile=default` for installing default profile, `profile=demo` for installing demo profile, and `profile=none` for not installing istio at all. This line can be placed anywhere before the cleanup steps.
 
-Your test script can then invoke the commands by simply calling snip functions:
+After setup, you will use snippets generated from the docs to write tests. The framework automatically sources several bash scripts for you, including the generated `snips.sh` and `tests/util/[verify|debug|helpers].sh`. You can directly call any function defined in them. For other test utilities, e.g., `util/samples.sh`, you need to source them by yourself:
```
```suggestion
After setup, you will use snippets generated from the docs to write tests. The framework automatically sources several bash scripts for you, including the generated `snips.sh` and `tests/util/[verify|debug|helpers].sh`. You can directly call any function defined in them. For other test utilities, e.g., `util/samples.sh`, you need to source them yourself:
```

frankbu commented on tests/README.md:
```diff
@@ -182,27 +148,45 @@ expected output. The framework includes the following built-in verify functions:
    for testing commands that demonstrate configurations that are expected to fail.
 
-## Running the Tests: Make
+After all test steps are run, add
```
```suggestion
After all test steps are complete, add the following line to indicate the start of the cleanup steps.
These steps will be run by framework, even if the test fails and prematurely exits.
```

frankbu commented on tests/README.md:
````diff
@@ -182,27 +148,45 @@ expected output. The framework includes the following built-in verify functions:
+```sh
+# @cleanup
+```
+on a single line, followed by the cleanup steps that clean up all the resources. These steps can also directly call functions defined in the auto-sourced scripts described before, as well as any script you have sourced by yourself for the test steps.
````
```suggestion
The following cleanup steps must remove all resources and reverse configuration changes made during the test steps. These steps can also directly call functions defined in the auto-sourced scripts described before, as well as any script you have sourced by yourself for the test steps.
```

frankbu commented on tests/README.md:
```diff
@@ -182,27 +148,45 @@ expected output. The framework includes the following built-in verify functions:
+You can also find this information by running `make doc.test.help`. The bash tracing output for debugging will be kept in `out/*_[test|cleanup]_debug.txt`.
+
+### Notes
+
+NOTE: The following were written for the old test framework and has NOT been tested on the new framework.
```
```suggestion
```

frankbu commented on tests/README.md:
```diff
@@ -214,16 +198,58 @@ make test.kube.trafficmanagement
-export REPO_ROOT=$(git rev-parse --show-toplevel)
-go test ./tests/... -p 1  --istio.test.env kube \
-    --istio.test.ci --istio.test.work_dir <my_dir>
+## Adding A Setup Configuration
```
This is advanced framework customization, not something we want to encourage test authors to do, so I don't think we want this in this README. How about moving it to another README in `tests/setup/`.

niwatorrri replied:
Sounds good!

frankbu commented on tests/README.md:
```diff
@@ -1,5 +1,7 @@
 # Testing istio.io Content
 
+**NOTE: We are now migrating to a new test framework. The following content is updated for the new framework.**
+
```
We are going to do a one time switch, no migrating, so you don't need to say anything about it in here.
```suggestion
```

frankbu commented on tests/README.md:
````diff
@@ -182,27 +148,45 @@ expected output. The framework includes the following built-in verify functions:
+```bash
+make doc.test TEST=traffic-management
+```
+will run all the tests under `traffic-management` folder. The `TEST` variable also accepts multiple test names separated by commas, for example,
````
run all tests under `traffic-managment` folder makes no sense. What `traffic-managment` folder? There are several. I think you should require a full path to the doc relative to `content/en/docs`. So it could be `TEST=tasks/traffic-managment` or `TEST=ops/configuration/traffic-managment` or `TEST=ops/best-practices/traffic-management` or other. Just giving the short name is confusing and not helpful.

niwatorrri replied:
How about [this](https://github.com/istio/istio.io/pull/7465/files#diff-62f0ff60ede3b78b9c95624e2f61d6c1R169)? I suggest it would be the recommended practice to specify full paths.

frankbu replied:
If you allow anything other than the full path, you need to explain how it works, and to be honest, I can't see a sensible answer. What happens if the user says "TEST=foo", or "TEST=foo/bar", or "TEST=foo/bar/baz"? Are you going to scan the entire repo looking for all parent directories that have the specified subdir(s)? It seems like a big complication for a small gain. Users will typically be using this feature to test a doc that they know the full path for, so copy and pasting that after the `TEST=` part of the command is pretty easy and straightforward.

niwatorrri replied:
The current way of matching tests is to traverse through the entire folder and check if a path contains ${TEST} as a substring, so ``foo/bar/baz`` will match short paths like `foo`, `bar`, `foo/bar` and also the full path `foo/bar/baz`. Okay I see this is kind of confusing once I explain it... let me change that.

niwatorrri replied:
This is now done by the latest commit. Take a look!

frankbu commented:
> Are we assuming only English docs will be tested? If that's the case yeah I can make them short.

Yes we only test English docs because the other languages would all be the exact same test, since they are just translations of the surrounding text, while running the exact same commands.

nmittler commented:
Woot! Great work, @niwatorrri! :)
